source/common/upstream/health_checker_impl.cc

```diff
@@ -263,6 +264,19 @@ HttpHealthCheckerImpl::HttpActiveHealthCheckSession::healthCheckResult() {
 
   const auto degraded = response_headers_->EnvoyDegraded() != nullptr;
 
+  if (parent_.service_name_matcher_feature_.enabled() && parent_.has_service_name_matcher_) {
```

Comment:
The way I would implement this is to load the deprecated field (if present) into a prefix string matcher, and then just have a single code path, then you can use the existing kill switch per my previous comment.